### PR TITLE
Add production hybrid latent MoE configs, EMO optimizations and HF export

### PR DESCRIPTION
**Pull request by jacob-morrison**
## Summary

Extract the reusable OLMo 3.5 hybrid latent-MoE production work onto current main,
without merging the experiment/campaign branch. One commit per logical chunk:

1. Per-head Q/K norm gains and scalable softmax.
2. EMO routing/global load balancing and opt-in document-pool/top-16 paths.
3. Accepted expert backward/FP32 accumulation/reduce-scatter optimizations;
   EP prewarm/ownership guards and LC recomputation compatibility.
4. Lossless HF state mapping and hybrid modeling, including legacy config migration.
5. Dependency-free checkpoint-ready notifications.
6. Portable model/training/data/retention recipe builders and review documentation.

Start review at **src/examples/olmo_ddp/OLMO3P5.md**. Small/medium/large geometry,
active/total counts and per-head gains are guarded. EMO selection is explicit;
the intended downstream MT/LC/SFT stages disable it.

## Scope boundaries

- Preserve current main's dependency pins and kernel-fun API. No vLLM fork.
- Optimizations remain opt-in. Historical 100B gains are not claimed for this rebase.
- No campaign job IDs, private mount paths, cluster launchers, watcher rollout,
  rejected communication probes, async checkpointing or precision changes.
- Uploader stays independent; the callback only publishes atomic filesystem events.
  Nothing in core uploads or deletes checkpoint data.
- Trainer retention explicitly disables main's default three-checkpoint cap in this
  recipe, so uploader verification remains the deletion gate.

## Important draft/requalification items

- Custom top-16 tie order is qualified only for Torch 2.11; the recipe selects
  native top-k on current Torch instead of widening that guard without evidence.
- Rounded wgrad requires QuACK 0.5.0 / SM100; current dependency/hardware combinations
  and EP8 need fresh GPU parity and save/resume tests.
- Medium gradient-norm differences remain an investigation/sign-off item.
- Medium CBS LR is provisional, not a 14T hero LR. Large production topology/LR
  are deliberately not invented.
- HF tensor export correctness is distinct from fast runtime numerical parity.

## Validation

- Changed core test files: **167 passed, 1255 skipped** on CPU.
  Skips include CUDA/multi-GPU/optional-kernel tests; this is not full GPU qualification.
- Recipe smoke/guards: **7 passed** (plus the repaired CPU gradient test in its own run).
- All three model parameter counts and actual Dolma3.5 manifest interpolation/
  ordered-path hash checked; 1167 data files.
- Python compilation, repo-pinned Ruff lint, Black and isort checks on touched files.
- No training/evaluation jobs launched and no live branch/service changed for this PR.

CI follow-up: repaired the normalized-attention default path, retained explicit
rejection of unsupported hybrid flags there, made router tests honor the
Torch-2.11-only top-16 contract, repaired test/type annotations and updated the
changelog. Local full-repo mypy passes; 12 targeted CPU regression checks pass
(one additional GPU case skips). These fixes are folded into their logical
commits; fresh CI is pending. GPU CI is not a replacement for production-scale
requalification of the new-main/dependency combination.

## Companion work and provenance

- Uploader: https://github.com/allenai/olmo-checkpoint-uploader/pull/1
- Inference plugins: https://github.com/allenai/scaling-ladders/pull/70
- Production source d031ab975c0dd11e986b3f017e0aac2e3608b521;
  accepted bundle 107dfa3ff42f4ee4984d5c445d587ceb7db5e4f4.
- EMO integration originates from the existing akshitab EMO work; KDA kernels
  remain upstream in kernel-fun. This is a curated extraction, not a claim that
  all pre-existing functionality was authored here.

### CODE REVIEW

File: src/examples/olmo_ddp/OLMO3P5.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
+# OLMo 3.5 hybrid latent MoE: reusable production recipe
```
Comment (AkshitaB) [inline src/examples/olmo_ddp/OLMO3P5.md]: We can remove this, since it's experimental

File: src/examples/olmo_ddp/olmo3p5_models.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
+"""Canonical small, medium and large OLMo 3.5 hybrid latent MoE shapes.
+
```
Comment (AkshitaB) [inline src/examples/olmo_ddp/olmo3p5_models.py]: Remove; experimental

File: src/examples/olmo_ddp/olmo3p5_recipe.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
+"""Portable OLMo 3.5 recipes, independent of Beaker campaigns or private paths.
+
```
Comment (AkshitaB) [inline src/examples/olmo_ddp/olmo3p5_recipe.py]: Remove; experimental

File: src/examples/olmo_ddp/test_olmo3p5_recipe.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+"""CPU-only drift guards for the reusable production recipe."""
```
Comment (AkshitaB) [inline src/examples/olmo_ddp/test_olmo3p5_recipe.py]: Remove; experimental

File: src/olmo_core/nn/attention/__init__.py [hunk 5]
```diff
@@ -482,6 +494,10 @@ def build(
         if sliding_window_config is not None and sliding_window_config.should_use_swa(
             layer_idx, n_layers
         ):
+            if self.scalable_softmax:
+                raise OLMoConfigurationError(
```
Comment (AkshitaB) [inline src/olmo_core/nn/attention/__init__.py:499]: Should also add warning about CP support

File: src/olmo_core/nn/hf/config.py [hunk 6]
```diff
@@ -333,6 +336,214 @@ def _get_olmo3moe_config(model: "OLMoDDPModel") -> PretrainedConfig:
     )
 
 
+def _get_olmo3moe_kda_emo_config(model: "OLMoDDPModel") -> PretrainedConfig:
+    """Build a fail-closed HF config for the KDA + full-width EMo ladder."""
```
Comment (AkshitaB) [inline src/olmo_core/nn/hf/config.py]:
```suggestion
    """Build a fail-closed HF config for the KDA + full-width EMo model."""
```

File: src/olmo_core/ops/moe_top16.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
+"""Exact-shape Torch 2.11 CUDA top16 index selection; default-off profiling use.
```
Comment (AkshitaB) [inline src/olmo_core/ops/moe_top16.py]: main has moved up to torch 2.13; confirm that it works with updated version

File: src/olmo_core/distributed/checkpoint/utils.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
+"""Strict, opt-in shared-to-per-head QK gain migration; never modifies the source checkpoint."""
```
Comment (AkshitaB) [inline src/olmo_core/distributed/checkpoint/utils.py:1]: Move to utils?

**Review by AkshitaB — CHANGES_REQUESTED** [review]

**Comment by jacob-morrison** [comment]
Addressed the review in 2f84f53596f3987f7a40fc01575af8d18d91c8bc: removed the four experimental recipe files, documented and guarded CP/KV limitations, moved the checkpoint helper to `distributed/checkpoint/utils.py`, and qualified top-16 plus the retained optimizations on Torch 2.13.

Also fixed legacy fused-attention construction and the latent/non-EMO/per-head/scalable-softmax HF export gaps. New tests build actual core models and check export, reload, forward agreement, and FP32 expert combination against core's TE unpermute.

The paired SwiGLU backward kernel failed exact-gradient qualification on Torch 2.13, so it and the unqualified rounded-wgrad path are deferred from this PR. The description now reflects the final scope and validation. Local checks pass; full repository CI is running.

**Comment by AkshitaB** [comment]
Some more comments from codex:

1. [P2] Hybrid export silently replaces sliding-window attention with full attention. config.py:536–539
     (https://github.com/allenai/OLMo-core/blob/f38332520f78cc6c71c6e21461a586fc90beb8c2/src/olmo_core/nn/hf/config.py#L536-L539) labels every non-KDA
     layer "full_attention" without checking its backend window. A hybrid model with sliding attention and scalable_softmax=False passes the export guards
     but gains access to tokens outside its original window. Tensor round-trip checks cannot detect this. Preserve the window configuration or reject these
     models explicitly.

  2. [P2] Non-KDA EMO export silently drops document-pool restrictions. config.py:260
     (https://github.com/allenai/OLMo-core/blob/f38332520f78cc6c71c6e21461a586fc90beb8c2/src/olmo_core/nn/hf/config.py#L260) accepts EMO routers without
     checking their evaluation pool size. The restricted-pool guard exists only in the KDA export path. Consequently, an attention-only EMO model
     evaluating with a pool of 2 out of 4 experts exports to an HF router selecting across all 4, changing expert assignments. Apply the evaluation-pool
     guard to both export paths.
